### PR TITLE
docs: added out-of-order and out-of-bound receiver troubleshoot cases;

### PR DESCRIPTION
**Pull request by ethernoy**
* [X] I added CHANGELOG entry for this change.
* [X] Change is not relevant to the end user.

## Changes

- Added Receiver section in troubleshoot document.
- Added troubleshoot case for out-of-bound Receiver error.
- Added troubleshoot case for out-of-order Receiver error.

## Verification

- Tested the solution in my own cluster

### CODE REVIEW

File: docs/operating/troubleshooting.md [hunk 1]
```diff
@@ -78,3 +78,74 @@ global:
...
+### Root Cause
+
+- TSDB expects to write samples in chronological order for each series.
+- A sample with timestamp t1 is sent to the Thanos Receiver and accepted, any sample with **timestamp t < t1** and **identical label set** being sent to the receiver after this will be determined as out-of-order sample.
+
+### Possible Cause
+
+- Remote Prometheus is running in high availability mode (more than 1 replica are running). But the replica_external_label_name is not correctly configured (e.g. empty).
+
+![](../img/thanos_receiver_troubleshoot_empty_replica_external_label_name.png)
```
Comment (wiardvanrij) [inline docs/operating/troubleshooting.md:129]: Please add alt text to these images with a description of the image. For example:

```
![Example view of compactor not coping with amount and size of incoming blocks](compactor_no_coping_with_load.png)
``` 
This improves accessibility :)

**Comment by wiardvanrij** [comment]
This is really awesome! Love it. 
I checked the preview (https://deploy-preview-4134--thanos-io.netlify.app/tip/operating/troubleshooting.md/#out-of-bound-error) and I think the images are a bit out of bounds. 

I actually have no idea how to fix this other than scaling the images ourself to a certain max-width. (I will ask this on Slack if anyone got an idea) 
Other than that, LGTM!

**Comment by ethernoy** [comment]
> This is really awesome! Love it.
> I checked the preview (https://deploy-preview-4134--thanos-io.netlify.app/tip/operating/troubleshooting.md/#out-of-bound-error) and I think the images are a bit out of bounds.
> 
> I actually have no idea how to fix this other than scaling the images ourself to a certain max-width. (I will ask this on Slack if anyone got an idea)
> Other than that, LGTM!

Thanks! I will resolve the accessibility issue tonight (I'm in UTC+8). Please let me know if you have any suggestions on how to resize the image from markdown point of view, if it can't be resolved here, I will consider resizing the image file. By the way, do you think it is a good idea to put the source .drawio file of those diagrams in the repository?

**Comment by ethernoy** [comment]
> This is really awesome! Love it.
> I checked the preview (https://deploy-preview-4134--thanos-io.netlify.app/tip/operating/troubleshooting.md/#out-of-bound-error) and I think the images are a bit out of bounds.
> 
> I actually have no idea how to fix this other than scaling the images ourself to a certain max-width. (I will ask this on Slack if anyone got an idea)
> Other than that, LGTM!

Hi, I updated the PR:

- added several alt text to images for better accessibility
- added fluid-img css class to images to improve the style of the page (now it won't go off-border)

**Review by wiardvanrij — DISMISSED** [review]
LGTM, really neat.

File: CHANGELOG.md [hunk 2]
```diff
@@ -15,6 +15,7 @@ We use _breaking :warning:_ to mark changes that are not backward compatible (re
 ### Added
 
 - [#4107](https://github.com/thanos-io/thanos/pull/4107) Store: `LabelNames` and `LabelValues` now support label matchers.
+- [#4134](https://github.com/thanos-io/thanos/pull/4134) Docs: add out-of-order and out-of-bound receiver troubleshoot cases.
```
[the remark this answers is not in the document]
Comment (ethernoy) [reply CHANGELOG.md:18]: sure, will remove it in next commit

File: docs/operating/troubleshooting.md [hunk 3]
```diff
@@ -78,3 +78,74 @@ global:
...
+```
+
+### Root Cause
+
+- "Out-of-bound" error occurs when the timestamp of the to-be-written sample is lower than the minimum acceptable timestamp of the TSDB head.
+
+### Possible Cause
+
+1. Thanos Receiver was stopped previously and is just resumed, remote Prometheus starts to write from the oldest sample, which is too old to be digested and hence rejected.
+1. Thanos Receiver does not have enough resources to ingest the remote write data. The latest ingested sample is gradually falling behind the latest scraped samples.
```
[the remark this answers is not in the document]
Comment (ethernoy) [reply docs/operating/troubleshooting.md]: > Can we also add a case from your amazing diagrams? (: (Multiple receivers adding sample for the same series without unique external labels (e.g replica labels).

I can add a diagram, but need more information for this case. Do you mean there will be `Error on ingesting samples that are too old or are too far into the future` when:


- there is one Prometheus remote writing to multiple receivers
- those receivers does not configure external_label
- those receivers does not configure a hashring (thus request for the same serie are handled by different receivers)

**Review by bwplotka — DISMISSED** [review]
Solid and super helpful. Thank you!

File: docs/operating/troubleshooting.md [hunk 4]
```diff
@@ -78,3 +78,74 @@ global:
...
+### Root Cause
+
+- TSDB expects to write samples in chronological order for each series.
+- A sample with timestamp t1 is sent to the Thanos Receiver and accepted, any sample with **timestamp t < t1** and **identical label set** being sent to the receiver after this will be determined as out-of-order sample.
+
+### Possible Cause
+
+- Remote Prometheus is running in high availability mode (more than 1 replica are running). But the replica_external_label_name is not correctly configured (e.g. empty).
+
+<img src="../img/thanos_receiver_troubleshoot_empty_replica_external_label_name.png" class="img-fluid" alt="Example topology diagram of out-of-order error case caused by empty replica_external_label_name" />
```
Comment (bwplotka) [inline docs/operating/troubleshooting.md:129]: BTW, how you created those diagrams? If Google diagrams, do you mind linking the Google drawing and sharing edit rights with Thanos maintainers for easier editions? (:

Comment (bwplotka) [reply docs/operating/troubleshooting.md:129]: you can use our email `thanos-io@googlegroups.com`

Comment (ethernoy) [reply docs/operating/troubleshooting.md:129]: > BTW, how you created those diagrams? If Google diagrams, do you mind linking the Google drawing and sharing edit rights with Thanos maintainers for easier editions? (:

I used draw.io, the source files are in xml format. I don't mind sharing those with you guys, maybe I can just sent those source files to the email group in attachment?

Comment (wiardvanrij) [reply docs/operating/troubleshooting.md:129]: Maybe add drawio as SVG? WIth https://marketplace.visualstudio.com/items?itemName=hediet.vscode-drawio we can then just IN EDITOR edit svg's. Which contains BOTH the drawio base AND the image itself. 
Pretty..neat with some overhead of-course

Comment (ethernoy) [reply docs/operating/troubleshooting.md:129]: > Maybe add drawio as SVG? WIth https://marketplace.visualstudio.com/items?itemName=hediet.vscode-drawio we can then just IN EDITOR edit svg's. Which contains BOTH the drawio base AND the image itself.
> Pretty..neat with some overhead of-course

I can save those drawIO files as svg/png (with drawIO source inside, which is its feature) in the thanos repository and refer to them directly in the troubleshoot document, will do it ASAP.

Comment (ethernoy) [reply docs/operating/troubleshooting.md]: added

File: docs/operating/troubleshooting.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (bwplotka) [reply docs/operating/troubleshooting.md]: I think I missed the error, I meant the other error case, so all good!

**Comment by bwplotka** [comment]
Thank you!

**Review by bwplotka — APPROVED** [review]

**Pull request merged**